core/runtime/src/text/mod.rs

```diff
@@ -0,0 +1,150 @@
+//! Module implementing JavaScript classes to handle text encoding and decoding.
+//!
+//! See <https://developer.mozilla.org/en-US/docs/Web/API/Encoding_API> for more information.
+
+use boa_engine::object::builtins::{JsArrayBuffer, JsTypedArray, JsUint8Array};
+use boa_engine::realm::Realm;
+use boa_engine::value::TryFromJs;
+use boa_engine::{
+    Context, Finalize, JsData, JsObject, JsResult, JsString, JsValue, Trace, js_error, js_string,
+};
+use boa_interop::boa_macros::{boa_class, boa_module};
+use std::rc::Rc;
+
+#[cfg(test)]
+mod tests;
+
+mod encodings;
+
+/// The [`TextDecoder`][mdn] class represents an encoder for a specific method, that is
+/// a specific character encoding, like `utf-8`.
+///
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder
+#[derive(Debug, Clone, JsData, Trace, Finalize)]
+pub struct TextDecoder(#[unsafe_ignore_trace] Rc<dyn encodings::Encoder>);
```

Comment:
This seems a bit overengineered IMO. What is the problem with using a simple enum with the list of all encodings we support? Should be smaller and faster than dynamic objects.

Reply:
So there is a point where I'll want to add a lot of encoders (using https://crates.io/crates/encoding_rs), but I guess that's not necessary to have a vtable for it. At least not yet.

Reply:
More so because the encoders don't carry state, they're just plain old algorithms